command/agent/monitor/monitor.go

```diff
@@ -0,0 +1,117 @@
+package monitor
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+// Monitor provides a mechanism to stream logs using go-hclog
+// InterceptLogger and SinkAdapter. It allows streaming of logs
+// at a different log level than what is set on the logger.
+type Monitor struct {
+	sync.Mutex
+	sink            log.SinkAdapter
+	logger          log.InterceptLogger
+	logCh           chan []byte
+	droppedCount    int
+	bufSize         int
+	droppedDuration time.Duration
+}
+
+// New creates a new Monitor. Start must be called in order to actually start
+// streaming logs
+func New(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) *Monitor {
+	sw := &Monitor{
+		logger:          logger,
+		logCh:           make(chan []byte, buf),
+		bufSize:         buf,
+		droppedDuration: 3 * time.Second,
+	}
+
+	opts.Output = sw
+	sink := log.NewSinkAdapter(opts)
+	sw.sink = sink
+
+	return sw
+}
+
+// Start registers a sink on the monitors logger and starts sending
+// received log messages over the returned channel. A non-nil
+// sopCh can be used to deregister the sink and stop log streaming
+func (d *Monitor) Start(stopCh <-chan struct{}) <-chan []byte {
+	d.logger.RegisterSink(d.sink)
+
+	streamCh := make(chan []byte, d.bufSize)
+	go func() {
+		defer close(streamCh)
+		for {
+			select {
+			case log := <-d.logCh:
+				select {
+				case <-stopCh:
+					d.logger.DeregisterSink(d.sink)
+					close(d.logCh)
+					return
+				case streamCh <- log:
+				}
+			case <-stopCh:
+				d.Lock()
+				defer d.Unlock()
```

Comment:
not sure whether this lock/unlock is needed, but it doesn't hurt anything.

Reply:
I think its unnecessary as well, will remove